### PR TITLE
Enable an optional namespace parameter for hasAction & hasFilter

### LINKED ISSUE #7819: TRY: Support an optional namespace parameter for hasAction/hasFilter
@adamsilverstein commented on [Wed Apr 11 2018](https://github.com/WordPress/packages/pull/106)

* Add tests
* Update docblock

Fixes https://github.com/WordPress/packages/issues/104

---

@codecov[bot] commented on [Tue Apr 24 2018](https://github.com/WordPress/packages/pull/106#issuecomment-380240851)

# [Codecov](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=h1) Report
> Merging [#106](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=desc) into [master](https://codecov.io/gh/WordPress/packages/commit/21f3ceae08e345b6de213165331278c568a83cc1?src=pr&el=desc) will **increase** coverage by `0.65%`.
> The diff coverage is `100%`.

[Impacted file tree graph](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #106      +/-   ##
==========================================
+ Coverage   64.07%   64.72%   +0.65%     
==========================================
  Files          42       42              
  Lines         604      618      +14     
  Branches      118      121       +3     
==========================================
+ Hits          387      400      +13     
- Misses        175      176       +1     
  Partials       42       42
```

| [Impacted Files](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [packages/hooks/src/createHasHook.js](https://codecov.io/gh/WordPress/packages/pull/106/diff?src=pr&el=tree#diff-cGFja2FnZXMvaG9va3Mvc3JjL2NyZWF0ZUhhc0hvb2suanM=) | `100% <100%> (ø)` | :arrow_up: |
| [packages/i18n/src/index.js](https://codecov.io/gh/WordPress/packages/pull/106/diff?src=pr&el=tree#diff-cGFja2FnZXMvaTE4bi9zcmMvaW5kZXguanM=) | `93.54% <0%> (-1.46%)` | :arrow_down: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=footer). Last update [21f3cea...e928084](https://codecov.io/gh/WordPress/packages/pull/106?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

---

@adamsilverstein commented on [Tue Jun 26 2018](https://github.com/WordPress/packages/pull/106#issuecomment-400289185)

@aduth any concerns for merging this, seems useful?

---

@adamsilverstein commented on [Fri Jun 29 2018](https://github.com/WordPress/packages/pull/106#issuecomment-401167576)

@aduth thanks for the feedback, I'll work on these soon

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Description
Core originally in from https://github.com/WordPress/packages/pull/106/
Fixes https://github.com/WordPress/gutenberg/issues/7819

## How has this been tested?
* Add a hook with its required namespace.
* Call matching `hasfilter/hasAction` and verify results correctly.
* testing calling without namespace continues to work as expected.
* See test files for various tests.

## Types of changes
* Add an optional namespace parameter for `hasfilter` & `hasAction` 
* Bdetter match core php functions`has_action` `has_filter` which include this second parameter

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/hooks/src/createHasHook.js [hunk 1]
```diff
@@ -5,17 +5,24 @@
...
+	 * @param {string} hookName  The name of the hook to check for.
+	 * @param {string} namespace Optional. The unique namespace identifying the callback in the form `vendor/plugin/function`.
 	 *
 	 * @return {boolean} Whether there are handlers that are attached to the given hook.
 	 */
-	return function hasHook( hookName ) {
+	return function hasHook( hookName, namespace ) {
+		// Use the namespace if provided.
+		if ( 'undefined' !== typeof namespace ) {
+			return hookName in hooks &&
```
Comment (aduth) [inline packages/hooks/src/createHasHook.js]: If `hookName in hooks` is common to both the `true` and `false` cases of the condition, could it be extracted out as an initial test?

Additionally, in pursuit of shorter line length, we could invert the condition to allow this test to occur two tabs up:

```js
if ( ! ( hookName in hooks ) ) {
	return false;
}

if ( typeof namespace === 'undefined' ) {
	return true;
}

return hooks[ hookName ].handlers.some( ( hook ) => hook.namespace === namespace );
```

Thoughts? I could see the `return true;` being a bit more confusing than as proposed, though perhaps easy enough to clarify with an inline code comment.

**Review by aduth — APPROVED** [review]

**Comment by aduth** [comment]
There's a merge conflict to be resolved as well.

**Comment by aduth** [comment]
Can we include a CHANGELOG.md note? I assume it should be publicly documented as well in README.md. 

Separately, it would be good to see what it would take to automate the documentation using [`docgen`](https://github.com/WordPress/gutenberg/tree/master/packages/docgen) like in other packages, so we don't need to maintain this documentation manually.

File: packages/hooks/src/createHasHook.js [hunk 2]
```diff
@@ -5,17 +5,24 @@
...
+ *                          attached to a particular hook and optional namespace.
  */
 function createHasHook( hooks ) {
 	/**
-	 * Returns how many handlers are attached for the given hook.
+	 * Returns whether any handlers are attached for the given hookName and optional namespace.
 	 *
-	 * @param  {string}  hookName The name of the hook to check for.
+	 * @param {string} hookName  The name of the hook to check for.
+	 * @param {string} namespace Optional. The unique namespace identifying the callback in the form `vendor/plugin/function`.
```
Comment (swissspidy) [inline packages/hooks/src/createHasHook.js]:
```suggestion
	 * @param {?string} namespace Optional. The unique namespace identifying the callback in the form `vendor/plugin/function`.
```

File: packages/hooks/src/createHasHook.js (hunk 1 above)
Comment (adamsilverstein) [reply packages/hooks/src/createHasHook.js]: The current code feels clearer and is more concise. If the line length is a concer, can we split the line instead?

**Comment by adamsilverstein** [comment]
@gziolo This could use a review if you have time. I'm not sure why there are failing checks.

File: packages/hooks/src/test/index.test.js [hunk 3]
```diff
@@ -740,3 +740,40 @@ test( 'add multiple all actions and run it any hook to trigger them by priority'
...
+	// hasAction tests
+	addAction( 'test.action', 'my_callback', () => {} );
+	expect( hasAction( 'test.action', 'not_my_callback' ) ).toBe( false );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( true );
+
+	// test removing
+	removeAction( 'test.action', 'my_callback' );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( false );
+
+	// test removeAll
```
Comment (gziolo) [inline packages/hooks/src/test/index.test.js]: This looks like it can be a start on its own `test`.

File: packages/hooks/src/test/index.test.js [hunk 4]
```diff
@@ -740,3 +740,40 @@ test( 'add multiple all actions and run it any hook to trigger them by priority'
...
+	addAction( 'test.action', 'my_callback', () => {} );
+	addAction( 'test.action', 'my_second_callback', () => {} );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( true );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( true );
+	removeAllActions( 'test.action' );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( false );
+	expect( hasAction( 'test.action', 'my_callback' ) ).toBe( false );
+
+	// hasFilter tests
```
Comment (gziolo) [inline packages/hooks/src/test/index.test.js]: Yet another test could start here.

File: packages/hooks/src/test/index.test.js [hunk 5]
```diff
@@ -740,3 +740,40 @@ test( 'add multiple all actions and run it any hook to trigger them by priority'
...
+	addFilter( 'test.filter', 'my_callback', () => {} );
+	expect( hasFilter( 'test.filter', 'not_my_callback' ) ).toBe( false );
+	expect( hasFilter( 'test.filter', 'my_callback' ) ).toBe( true );
+
+	// test removing
+	removeFilter( 'test.filter', 'my_callback' );
+	expect( hasFilter( 'test.filter', 'my_callback' ) ).toBe( false );
+
+	// test removeAll
+	addFilter( 'test.filter', 'my_callback', () => {} );
```
Comment (gziolo) [inline packages/hooks/src/test/index.test.js]: Yet another one could start here :)

**Review by gziolo — APPROVED** [review]
I left only nitpicks :)

It looks good and I think we should also update API Usage section in README file:
https://github.com/WordPress/gutenberg/tree/master/packages/hooks#api-usage

File: packages/hooks/src/test/index.test.js (hunk 3 above)
Comment (adamsilverstein) [reply packages/hooks/src/test/index.test.js]: 👍

File: packages/hooks/src/test/index.test.js (hunk 4 above)
Comment (adamsilverstein) [reply packages/hooks/src/test/index.test.js]: 👍

**Comment by adamsilverstein** [comment]
@gziolo Thanks for the review. I will work on your feedback and update API Usage section in README.

**Comment by adamsilverstein** [comment]
@gziolo I addressed your feedback and this should be ready for a merge once all tests pass.

File: packages/hooks/README.md [hunk 6]
```diff
@@ -39,8 +39,8 @@ In the WordPress context, API functions can be called via the global `wp.hooks`
 * `doingFilter( 'hookName' )`
 * `didAction( 'hookName' )`
 * `didFilter( 'hookName' )`
-* `hasAction( 'hookName' )`
-* `hasFilter( 'hookName' )`
+* `hasAction( 'hookName', 'namespace )`
```
Comment (talldan) [inline packages/hooks/README.md]:
```suggestion
* `hasAction( 'hookName', 'namespace' )`
```

File: packages/hooks/README.md [hunk 7]
```diff
@@ -39,8 +39,8 @@ In the WordPress context, API functions can be called via the global `wp.hooks`
...
+* `hasFilter( 'hookName' 'namespace' )`
```
Comment (talldan) [inline packages/hooks/README.md]:
```suggestion
* `hasFilter( 'hookName', 'namespace' )`
```

**Comment by gziolo** [comment]
Let's get this PR in, @adamsilverstein thanks for all iterations applied.

**Pull request merged**